draft.md

```diff
@@ -0,0 +1,419 @@
+
+A fascinating aspect of Lisp-based programming languages is that [code is data and data is code](https://en.wikipedia.org/wiki/Code_as_data). This property, called [homoiconicity](https://en.wikipedia.org/wiki/Homoiconicity), is what makes Lisp macros so powerful. In Haskell, this sort of runtime operation on arbitrary datatypes (called [polytypism](https://en.wikipedia.org/wiki/Polymorphism_(computer_science)#Polytypism) or datatype genericity) is provided by generics, of which there are two varieties:
+
+1. [GHC Generics](https://hackage.haskell.org/package/base-4.16.0.0/docs/GHC-Generics.html)
+2. [generics-sop](https://hackage.haskell.org/package/generics-sop)
+
+While the former comes with `base`, the latter is much easier to write generics code in, and this blog post gives an introduction to using `generics-sop`.
+
+- [Generics SOP introduction](#generics-sop-introduction)
+  - [Motivation](#motivation)
+  - [Basics](#basics)
+    - [Datatypes are SOPs under the hood](#datatypes-are-sops-under-the-hood)
+    - [SOPs are tables](#sops-are-tables)
+    - [Interlude: a foray into type-level programming](#interlude-a-foray-into-type-level-programming)
+  - [Let's play with SOPs](#lets-play-with-sops)
+    - [Interlude: `NS` & `NP`](#interlude-ns--np)
+    - [Code as data; data as code](#code-as-data-data-as-code)
+  - [Example: generic equality](#example-generic-equality)
+    - [Naive implementation](#naive-implementation)
+    - [Combinators](#combinators)
+  - [Further information](#further-information)
+
+## Motivation
+
+Generic programming is useful in avoiding having to manually write some implementation for each datatype. This could be a ([polytypic](https://www.sciencedirect.com/science/article/pii/S0167642304000152)) function or a typeclass instance. For example, instead of having to manually write `FromJSON` and `ToJSON` instances for each of your datatypes, you can use generics to derive them automatically. Other examples include pretty printers, parsers, equality functions and route encoders.
+
+## Basics
+
+Before diving, we must understand the "SOP" in generics-**sop**.
+
+### Datatypes are SOPs under the hood
+
+Haskell has two kinds of datatypes:
+
+1. [Algebraic data types](https://en.wikipedia.org/wiki/Algebraic_data_type), or ADTs
+2. Newtypes
+
+Both of these can be converted to what is known as a "sum-of-product" (SOP). When writing generics-sop code, we operate on these SOPs rather than the datatype directly, because every datatype is "polymorphic" in their SOP representation. The basic idea is that if you can write a function `SOP -> a`, then you get `SomeDataType -> a` for free for *any* `SomeDataType`. This is called [polytypism](https://en.wikipedia.org/wiki/Polymorphism_(computer_science)#Polytypism).
+
+Consider the following ADT (from [the `these` package](https://hackage.haskell.org/package/these-1.1.1.1/docs/Data-These.html)):
+
+```haskell
+data These a b
+  = This a 
+  | That a
+  | These a b
```

Comment:
Should probably be `That b`, right? This also applies to the tables below.

Reply:
Thanks! Fixed.